internal/server/server.go

```diff
@@ -433,37 +463,95 @@ func (h *Handler) predictWithRunner(srcURL string, req PredictionRequest) (chan
+		// if "setupComplete" is closed, we know we didn't hit the timeout case.
+		//
+		// In theory this could be moved in the config .Go() run but for clarity we isolate since
+		// errors in .config() or below in the .Go() run below are both cases to .Stop() the runner.
+		select {
+		case <-setupComplete:
+			return nil
+		case <-deadlineCtx.Done():
+			log.Errorw("stopping procedure runner after timeout", "elapsed", time.Since(start))
+			// Stop the runner to avoid it hanging around, r.wait() cannot be canceled via context yet
+			// NOTE(morgan): failure here means we do not want to give a graceperiod to the runner
+			r.shutdownGracePeriod = 0
+			err := r.Stop()
+			if err != nil {
+				log.Errorw("failed to stop runner", "error", err)
 			}
-			if err := os.Remove(readyFile); err != nil && !os.IsNotExist(err) {
-				log.Errorw("failed to remove ready file", "error", err)
+			if errors.Is(deadlineCtx.Err(), context.DeadlineExceeded) {
+				return fmt.Errorf("%w: timeout waiting for runner to config", context.DeadlineExceeded)
 			}
-		}
-		// No ready file if runner dies but status should be setup failed
-		if r.status != StatusStarting {
-			break
-		}
-		if time.Since(start) > 10*time.Second {
-			log.Errorw("stopping procedure runner after time out", "elapsed", time.Since(start))
-			if err := r.Stop(); err != nil {
-				log.Errorw("failed to stop procedure runner", "error", err)
+			return deadlineCtx.Err()
+		}
+	})
+	eg.Go(func() error {
+		// Wait for runner to become ready, this should not take long as procedures have no setup
+		ticker := time.NewTicker(10 * time.Millisecond)
+		defer ticker.Stop()
+		for {
+			// We do not register non-ready runner yet for HTTP IPC due to potential race condition
+			// Instead we poll here for a ready file and status change
+			readyFile := path.Join(r.workingDir, "ready")
+			if _, err := os.Stat(readyFile); err == nil {
+				if err := r.HandleIPC(IPCStatusReady); err != nil {
+					log.Errorw("failed to handle IPC", "error", err)
+					return fmt.Errorf("failed to handle IPC: %w", err)
+				}
+				if err := os.Remove(readyFile); err != nil && !os.IsNotExist(err) {
+					log.Errorw("failed to remove ready file", "error", err)
+				}
+			}
+			// No ready file if runner dies but status should be setup failed
+			if r.status != StatusStarting {
+				// This is the ONLY goroutine tasked with closing the setupComplete channel
+				close(setupComplete)
+				return nil
+			}
+			select {
+			case <-ticker.C:
+				continue
+			case <-deadlineCtx.Done():
+				if errors.Is(deadlineCtx.Err(), context.DeadlineExceeded) {
+					return fmt.Errorf("%w: timeout waiting for runner to config", context.DeadlineExceeded)
+				}
+				return deadlineCtx.Err()
 			}
-			return nil, fmt.Errorf("procedure time out")
 		}
-		time.Sleep(10 * time.Millisecond)
+	})
+	// wait for the errorgroup to complete, we expect "setup completed" "setup failed" or "timeout"
```

Comment:
Just double checking the logic here since I think there's 2 "setup failed" cases that behave differently.
1) an unexpected error occurred in the setup function, which would cause the eg to get canceled and exit here. this is a Failure failure and we can't recover
2) an "expected" error occurred in the setup (eg user code failed) and the eg wait completes without error, allowing the next expression to detect `r.status` and return an error with logs.

Correct?